src/circuits/joinsplit.tcc

```diff
@@ -21,35 +22,43 @@ template<
 class joinsplit_gadget : libsnark::gadget<FieldT>
 {
 private:
-    // Multipacking gadgets for the inputs (nullifierS, commitmentS, val_pub_in,
-    // val_pub_out) (the root is a field element) `NumInputs + NumOutputs`
-    // because we pack the nullifiers (Inputs of JS = NumInputs), the
-    // commitments (Output of JS = NumOutputs) AND the v_pub_out taken out of
-    // the mix (+1) AND the public value v_pub_in that is put into the mix (+1)
-    // AND the signature hash h_sig (+1) AND the message authentication tags
-    // h_iS (+NumInputs)
+    // Number of residual bits from packing of hash digests into smaller
+    // field elements to which are added the public value of size 64 bits
+    const size_t length_bit_residual =
+        2 * 64 + (HashT::get_digest_len() - FieldT::capacity()) *
+                     (1 + 2 * NumInputs + NumOutputs);
+    // Number of field elements needed to pack this number of bits
+    const size_t nb_field_residual =
+        libff::div_ceil(length_bit_residual, FieldT::capacity());
+
+    // Multipacking gadgets for the inputs (nullifierS, commitmentS, hsig,
+    // message authentication tags (h_is) and the residual bits (comprising the
+    // previous variables' bits not containable in a single field element as
+    // well as the public values) (the root is a field element) because we pack
+    // the nullifiers (Inputs of JS = NumInputs), the commitments (Output of JS
+    // = NumOutputs) AND the signature hash h_sig (+1) AND the message
+    // authentication tags h_iS (+ NumInputs) AND the residual field elements
+    // which aggregate the extra bits and public values (+1)
     std::array<
         libsnark::pb_variable_array<FieldT>,
-        NumInputs + NumOutputs + 1 + 1 + 1 + NumInputs>
+        NumInputs + NumOutputs + 1 + NumInputs + 1>
```

Comment:
I thought it would be cleaner to have 1 pb_variable_array comprising `nb_field_residual` `FieldTs rather than the opposite.

`            // We allocate as many field elements as needed to pack the public
            // values and the hash digests' residual bits
            packed_inputs[NumInputs + NumOutputs + 1 + NumInputs].allocate(
                pb, nb_field_residual, FMT(this->annotation_prefix, " bits"));`

Reply:
I see, thanks